pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartialUpsertHandler.java

```diff
@@ -64,53 +62,42 @@ public PartialUpsertHandler(Schema schema, Map<String, UpsertConfig.Strategy> pa
    * For example, overwrite merger will only override the prev value if the new value is not null.
    * Null values will override existing values if not configured. They can be ignored by using ignoreMerger.
    *
-   * @param indexSegment the segment of the last derived full record during ingestion.
-   * @param docId the docId of the last derived full record during ingestion in the segment.
+   * @param prevRecord wrapper for previous record, which lazily reads column values of previous row and caches for
+   *                   re-reads.
    * @param newRecord the new consumed record.
    */
-  public void merge(IndexSegment indexSegment, int docId, GenericRow newRecord) {
-    for (String column : indexSegment.getColumnNames()) {
+  public void merge(LazyRow prevRecord, GenericRow newRecord) {
+    for (String column : prevRecord.getColumnNames()) {
       if (!_primaryKeyColumns.contains(column)) {
         PartialUpsertMerger merger = _column2Mergers.getOrDefault(column, _defaultPartialUpsertMerger);
         // Non-overwrite mergers
         // (1) If the value of the previous is null value, skip merging and use the new value
         // (2) Else If the value of new value is null, use the previous value (even for comparison columns).
         // (3) Else If the column is not a comparison column, we applied the merged value to it.
         if (!(merger instanceof OverwriteMerger)) {
-          try (PinotSegmentColumnReader pinotSegmentColumnReader = new PinotSegmentColumnReader(indexSegment, column)) {
-            if (!pinotSegmentColumnReader.isNull(docId)) {
-              Object previousValue = pinotSegmentColumnReader.getValue(docId);
-              if (newRecord.isNullValue(column)) {
-                // Note that we intentionally want to overwrite any previous _comparisonColumn value in the case of
-                // using
-                // multiple comparison columns. We never apply a merge function to it, rather we just take any/all
-                // non-null comparison column values from the previous record, and the sole non-null comparison column
-                // value from the new record.
-                newRecord.putValue(column, previousValue);
-                newRecord.removeNullValueField(column);
-              } else if (!_comparisonColumns.contains(column)) {
-                newRecord.putValue(column, merger.merge(previousValue, newRecord.getValue(column)));
-              }
+          if (!prevRecord.isNullValue(column)) {
```

Comment:
We can directly call `prevRecord.getValue(column)` and check if the `prevValue` is `null` to reduce one extra lookup

Reply:
^^